packages/js/product-editor/src/blocks/generic/conditional/edit.tsx

```diff
@@ -21,18 +23,25 @@ export function Edit( {
 	context,
 }: ProductEditorBlockEditProps< ConditionalBlockAttributes > ) {
 	const blockProps = useWooBlockProps( attributes );
-	const { mustMatch } = attributes;
+	const { mustMatch, showIf } = attributes;
 
-	const product = context.editedProduct;
+	if ( mustMatch ) {
+		deprecated( '`mustMatch` attribute in woocommerce/conditional block', {
+			alternative: '`showIf` attribute in woocommerce/conditional block',
+		} );
+	}
 
 	const displayBlocks = useMemo( () => {
-		for ( const [ prop, values ] of Object.entries( mustMatch ) ) {
-			if ( ! values.includes( product[ prop ] ) ) {
-				return false;
+		if ( mustMatch && context.editedProduct ) {
+			for ( const [ prop, values ] of Object.entries( mustMatch ) ) {
+				if ( ! values.includes( context.editedProduct[ prop ] ) ) {
```

Comment:
I am not a huge fan of passing the `editedProduct` through context, as this in turn will trigger the context to be updated on every product change. Meaning that even if a block wasn't using the `editedProduct` context, it would still be re-run, this can cause a drastic performance bottleneck in the future.
You can already see the difference between these two Performance snapshots ( see the scripting time and the CPU usage at the top, a much larger blocking time ) This is me just typing `test` in the name field:
With `editedProduct` passed to context:
<img width="1157" alt="Screenshot 2023-10-12 at 3 25 50 PM" src="https://github.com/woocommerce/woocommerce/assets/2240960/646a0c54-32e6-408a-a55a-377a7c7a3a87">

Without `editedProduct` passed to context:
<img width="1158" alt="Screenshot 2023-10-12 at 3 24 27 PM" src="https://github.com/woocommerce/woocommerce/assets/2240960/585ef346-ca1a-428f-8e27-112bc4808e75">

Reply:
An alternative is to just use `useSelect` in the conditional block, something like this:
```
const displayBlocks = useSelect( ( select ) => {
		const editedProduct = select( 'core' ).getEditedEntityRecord(
			'postType',
			productType,
			productId
		);
        if ( mustMatch && context.editedProduct ) {
			for ( const [ prop, values ] of Object.entries( mustMatch ) ) {
				if ( ! values.includes( context.editedProduct[ prop ] ) ) {
					return false;
				}
			}
		}

		return evaluate( showIf, context );
   }, [ mustMatch, showIf ]
);
```

There are a couple benefits of pushing this logic into the `useSelect`:
- The `useSelect` return value is memoized, so it won't trigger a re-render until the return value is updated.
- The `useSelect` inner function will still be run on every product change, but is independent of the React life cycle ( more performant ).
- The `useSelect` accepts the dependency array, so it will still re-trigger if `muchMatch` or `showIf` are passed in.

Reply:
@louwie17 Thanks for the review and for bringing up this performance concern. The plan for the conditional visibility (related to this) was based on having the edited product in the context for all blocks.

I've been re-working the implementation of the general conditional visibility based on your suggestion and I think I have a workable solution that is much more performant. I'll share the updated code in that PR later, and assuming that flies, I will also update this block to do something similar.